helper-scripts/backup_and_restore.sh

```diff
@@ -1,331 +1,621 @@
+      exit 1
+      ;;
+  esac
+done
+
+# Prevent passing --restore , --backup or --delete together
+declare -i OPTION_COUNT=0
+
+if [[ ! -z "${MAILCOW_DELETE_LOCATION}" ]]; then
+  OPTION_COUNT=$((OPTION_COUNT+1))
+fi
+
+if [[ ! -z "${MAILCOW_BACKUP_LOCATION}" ]]; then
+  OPTION_COUNT=$((OPTION_COUNT+1))
+fi
+
+if [[ ! -z "${MAILCOW_RESTORE_LOCATION}" ]]; then
+  OPTION_COUNT=$((OPTION_COUNT+1))
+fi
+
+if [[ "${OPTION_COUNT}" -gt 1 ]] || [[ "${OPTION_COUNT}" -eq 0 ]]; then
+  echo -e "${RED_COLOR}You should pass one of the following options: ${YELLOW_COLOR}-b/--backup${RESET}, ${YELLOW_COLOR}-r/--restore${RESET} or ${YELLOW_COLOR}-d/--delete${RESET}"
   exit 1
 fi
 
-echo "Using ${BACKUP_LOCATION} as backup/restore location."
-echo
+# Merge backup components if all is passed
+if [[ ! -z "${MAILCOW_BACKUP_LOCATION}" ]] || [[ ! -z "${MAILCOW_RESTORE_LOCATION}" ]]; then
+  if [[ " ${MAILCOW_BACKUP_COMPONENTS[*]} " =~ " all " ]]; then
+    # MAILCOW_BACKUP_COMPONENTS=("crypt" "vmail" "redis" "rspamd" "postfix" "mysql")
+    MAILCOW_BACKUP_COMPONENTS=("all")
+  fi
+fi
 
 source ${SCRIPT_DIR}/../mailcow.conf
 
 if [[ -z ${COMPOSE_PROJECT_NAME} ]]; then
-  echo "Could not determine compose project name"
+  echo -e "${RED_COLOR}Could not determine compose project name${RESET}"
   exit 1
 else
-  echo "Found project name ${COMPOSE_PROJECT_NAME}"
+  echo -e "${GREEN_COLOR}Found project name ${COMPOSE_PROJECT_NAME}${RESET}"
   CMPS_PRJ=$(echo ${COMPOSE_PROJECT_NAME} | tr -cd "[0-9A-Za-z-_]")
 fi
 
-if grep --help 2>&1 | head -n 1 | grep -q -i "busybox"; then
-  >&2 echo -e "\e[31mBusyBox grep detected on local system, please install GNU grep\e[0m"
-  exit 1
-fi
-
 
 function backup() {
   DATE=$(date +"%Y-%m-%d-%H-%M-%S")
-  mkdir -p "${BACKUP_LOCATION}/mailcow-${DATE}"
-  chmod 755 "${BACKUP_LOCATION}/mailcow-${DATE}"
-  cp "${SCRIPT_DIR}/../mailcow.conf" "${BACKUP_LOCATION}/mailcow-${DATE}"
-  touch "${BACKUP_LOCATION}/mailcow-${DATE}/.$ARCH"
-  for bin in docker; do
-  if [[ -z $(which ${bin}) ]]; then
-    >&2 echo -e "\e[31mCannot find ${bin} in local PATH, exiting...\e[0m"
-    exit 1
-  fi
-  done
+  mkdir -p "${MAILCOW_BACKUP_LOCATION}/mailcow-${DATE}"
+  chmod 755 "${MAILCOW_BACKUP_LOCATION}/mailcow-${DATE}"
+  cp "${SCRIPT_DIR}/../mailcow.conf" "${MAILCOW_BACKUP_LOCATION}/mailcow-${DATE}"
+  touch "${MAILCOW_BACKUP_LOCATION}/mailcow-${DATE}/.$ARCH"
+
+  echo -e "${GREEN_COLOR}Using ${MAILCOW_BACKUP_RESTORE_THREADS} thread(s) for this backup.${RESET}"
+
   while (( "$#" )); do
     case "$1" in
     vmail|all)
       docker run --name mailcow-backup --rm \
-        -v ${BACKUP_LOCATION}/mailcow-${DATE}:/backup:z \
+        -v "${MAILCOW_BACKUP_LOCATION}"/mailcow-${DATE}:/backup:z \
         -v $(docker volume ls -qf name=^${CMPS_PRJ}_vmail-vol-1$):/vmail:ro,z \
-        ${DEBIAN_DOCKER_IMAGE} /bin/tar --warning='no-file-ignored' --use-compress-program="pigz --rsyncable -p ${THREADS}" -Pcvpf /backup/backup_vmail.tar.gz /vmail
```

Comment:
tar is also needed to check in the required tools (line 30) then

pigz too i assume?

Reply:
Both `tar` and `pigz` are running from inside `mailcow/backup:latest` which is doing only one thing:

```Dockerfile
FROM debian:bookworm-slim

RUN apt update && apt install pigz
```

so no need to be checked in the host machine, right?

Reply:
Ah yes you're right... myself created this though :D No no need for action here.